distributed/dashboard/tests/test_scheduler_bokeh.py

```diff
@@ -610,6 +611,41 @@ async def test_TaskGraph_order(c, s, a, b):
     assert gp.node_source.data["state"][gp.layout.index[y.key]] == "erred"
 
 
+####Adding TGroupGraph tests ######
+@gen_cluster(client=True)
+async def test_TGroupGraph(c, s, a, b):
+    tgg = TGroupGraph(s)
+
+    futures = c.map(slowinc, range(10), delay=0.001)
+    await wait(futures)
+
+    tgg.update()
+    dnode = dict(tgg.node_source.data)
+    assert all(len(L) == 1 for L in dnode.values())
+    assert dnode["name"] == ["slowinc"]
+    assert dnode["tot_tasks"] == [10]
+
+    darr = dict(tgg.arrow_source.data)
+    assert all(len(L) == 0 for L in darr.values())
+
+    futures2 = c.map(dec, range(5))
+    await wait(futures2)
+
+    tgg.update()
+    dnode = dict(tgg.node_source.data)
+    assert all(len(L) == 2 for L in dnode.values())
+    assert dnode["name"] == ["dec", "slowinc"]  # they follow the tg_stack
+    assert dnode["tot_tasks"] == [5, 10]
+
+    del futures, futures2
```

Comment:
```suggestion
    del futures, futures2
    
    while s.tasks:
        await asyncio.sleep(0.01)
```

Reply:
@ncclementi probably the signal from the client that these futures have been removed hasn't yet reached the scheduler.  Let's wait here until that has occured, then we can proceed

Reply:
hmm. I've tried with a the wait there but it's still not clearing it, I even increase the sleep, but I can still see the source data.

Reply:
@ncclementi and I chatted about this a bit offline and it turned out we weren't updating the underlying source data when there were no `TaskGroup`s. We've pushed a commit which covers that case now and this test should be passing

Reply:
The test works now, but i just realized that it might not be deterministic at the moment since sometimes we have more than one Task group that has no dependencies (let's say `a` and `b`) which at the moment results in different layouts (see pictures below). When we compute the toposort order sometimes `a` is before `b` and sometimes it's the other way around. This can cause the test to fail since `tgg.node_source.data["name"] can be ["a", "b"] or ["b", "a"]. 

Either we don't test for the order, or we need to modify the logic on the toposort to do something extra when we have more than one node with zero dependencies. 

This computation can generate both graph layouts 

```python
from dask.distributed import Client
client = Client()

import dask.array as da
x = da.random.random((10000, 10000))
y = da.random.random((10000, 10000))
z = x.dot(y.T).persist()
```

![Screen Shot 2021-06-08 at 9 24 16 AM](https://user-images.githubusercontent.com/7526622/121399073-984c5d80-c945-11eb-89da-f841c2888c15.png)

![Screen Shot 2021-06-09 at 5 07 46 PM](https://user-images.githubusercontent.com/7526622/121399105-a1d5c580-c945-11eb-9f4f-2b2fb0cee33b.png)

Reply:
The ordering within the CDS doesn't matter, but the ordering on the page may.  I think that this requires us to have a conversation about what we actually want to visualize first.  I think that we should zoom out before we dive into making sure that everything here is running smoothly.